Include/cpython/pyatomic.h

```diff
@@ -0,0 +1,397 @@
+_Py_atomic_load_uint_relaxed(const unsigned int *address);
+
+static inline Py_ssize_t
+_Py_atomic_load_ssize_relaxed(const Py_ssize_t *address);
+
+static inline void *
+_Py_atomic_load_ptr_relaxed(const void *address);
+
+// Atomically performs `*address = value` (sequential consistency)
+static inline void
+_Py_atomic_store_int(int *address, int value);
+
+static inline void
+_Py_atomic_store_int8(int8_t *address, int8_t value);
+
+static inline void
+_Py_atomic_store_int16(int16_t *address, int16_t value);
+
+static inline void
+_Py_atomic_store_int32(int32_t *address, int32_t value);
+
+static inline void
+_Py_atomic_store_int64(int64_t *address, int64_t value);
+
+static inline void
+_Py_atomic_store_intptr(intptr_t *address, intptr_t value);
+
+static inline void
+_Py_atomic_store_uint8(uint8_t *address, uint8_t value);
+
+static inline void
+_Py_atomic_store_uint16(uint16_t *address, uint16_t value);
+
+static inline void
+_Py_atomic_store_uint32(uint32_t *address, uint32_t value);
+
+static inline void
+_Py_atomic_store_uint64(uint64_t *address, uint64_t value);
+
+static inline void
+_Py_atomic_store_uintptr(uintptr_t *address, uintptr_t value);
+
+static inline void
+_Py_atomic_store_uint(unsigned int *address, unsigned int value);
+
+static inline void
+_Py_atomic_store_ptr(void *address, void *value);
+
+static inline void
+_Py_atomic_store_ssize(Py_ssize_t* address, Py_ssize_t value);
+
+// Stores `*address = value` (relaxed consistency, i.e., no ordering)
+static inline void
+_Py_atomic_store_int_relaxed(int *address, int value);
+
+static inline void
+_Py_atomic_store_int8_relaxed(int8_t *address, int8_t value);
+
+static inline void
+_Py_atomic_store_int16_relaxed(int16_t *address, int16_t value);
+
+static inline void
+_Py_atomic_store_int32_relaxed(int32_t *address, int32_t value);
+
+static inline void
+_Py_atomic_store_int64_relaxed(int64_t *address, int64_t value);
+
+static inline void
+_Py_atomic_store_intptr_relaxed(intptr_t *address, intptr_t value);
+
+static inline void
+_Py_atomic_store_uint8_relaxed(uint8_t* address, uint8_t value);
+
+static inline void
+_Py_atomic_store_uint16_relaxed(uint16_t *address, uint16_t value);
+
+static inline void
+_Py_atomic_store_uint32_relaxed(uint32_t *address, uint32_t value);
+
+static inline void
+_Py_atomic_store_uint64_relaxed(uint64_t *address, uint64_t value);
+
+static inline void
+_Py_atomic_store_uintptr_relaxed(uintptr_t *address, uintptr_t value);
+
+static inline void
+_Py_atomic_store_uint_relaxed(unsigned int *address, unsigned int value);
+
+static inline void
+_Py_atomic_store_ptr_relaxed(void *address, void *value);
+
+static inline void
+_Py_atomic_store_ssize_relaxed(Py_ssize_t *address, Py_ssize_t value);
+
+// Stores `*address = value` (release operation)
+static inline void
+_Py_atomic_store_uint64_release(uint64_t *address, uint64_t value);
+
+static inline void
+_Py_atomic_store_ptr_release(void *address, void *value);
+
+
+// Sequential consistency fence
```

Comment:
I would prefer a more elaborated documentation, "fence" is kind of weak.

[__faststorefence doc](https://learn.microsoft.com/en-us/cpp/intrinsics/faststorefence?view=msvc-170):

> Guarantees that every previous memory reference, including both load and store memory references, is globally visible before any subsequent memory reference.

[_ReadWriteBarrier doc](https://learn.microsoft.com/en-us/cpp/intrinsics/readwritebarrier?view=msvc-170):

> Limits the compiler optimizations that can reorder memory accesses across the point of the call.

[dmb ish](https://devblogs.microsoft.com/oldnewthing/20220812-00/?p=106968):

> The data memory barrier ensures that all preceding writes are issued before any subsequent memory operations (including speculative memory access).

Reply:
The challenge I have is that it's really hard to describe what fences do in a way that's helpful and accurate. The above documentation is too strong for C11 fences.

There's https://en.cppreference.com/w/c/atomic/atomic_thread_fence, but I find it vague. And the C++ documentation (https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence) is more detailed but really hard to understand.

Reply:
Please add one or two of these links here. It's ok to have references to external doc, it's better than no doc :-)